File: go/leetcode/328. Odd Even Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func oddEvenList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {return head}
    start := 1
    nav := head
    odd_start := &ListNode{0, nil}
    odd_end := odd_start
    even_start := &ListNode{9, nil}
    even_end := even_start 
    
    for nav != nil {
        if start % 2 == 0 {
            // even
            even_end.Next = nav
            nav = nav.Next
            even_end = even_end.Next
            even_end.Next = nil
            start++ 
            
        } else {
            // odd
            odd_end.Next = nav
            nav = nav.Next
            odd_end = odd_end.Next
            odd_end.Next = nil
            start++
        }
    } 
        
    odd_end.Next = even_start.Next
    return odd_start.Next
}

File: c++/leetcode/328. Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
  ListNode *oddEvenList(ListNode *head)
  {
    if (!head || !head->next)
      return head;
    int start = 1;
    ListNode *nav = head;

    ListNode *odd_start = new ListNode;
    ListNode *odd_end = odd_start;
    ListNode *even_start = new ListNode;
    ListNode *even_end = even_start;

    while (nav)
    {
      if (start % 2 == 0)
      {
        // even
        even_end->next = nav;
        nav = nav->next;
        even_end = even_end->next;
        even_end->next = nullptr;
        start++;
      }
      else
      {
        // odd
        odd_end->next = nav;
        nav = nav->next;
        odd_end = odd_end->next;
        odd_end->next = nullptr;
        start++;
      }
    }

    // connect nodes
    odd_end->next = even_start->next;
    odd_start = odd_start->next;

    // return connected nodes
    return odd_start;
  }
};